platform/viewer/src/routes/Mode/Mode.tsx

```diff
@@ -38,20 +38,41 @@ async function defaultRouteInit({
 
   unsubscriptions.push(instanceAddedUnsubscribe);
 
-  const { unsubscribe: seriesAddedUnsubscribe } = DicomMetadataStore.subscribe(
-    DicomMetadataStore.EVENTS.SERIES_ADDED,
-    ({ StudyInstanceUID, madeInClient }) => {
-      const studyMetadata = DicomMetadataStore.getStudy(StudyInstanceUID);
-      if (!madeInClient) {
-        HangingProtocolService.run(studyMetadata);
-      }
+  const { unsubscribe: seriesAddedUnsubscribe } = DisplaySetService.subscribe(
+    DisplaySetService.EVENTS.DISPLAY_SETS_CHANGED,
+    displaySets => {
+      if (!displaySets || !displaySets.length) return;
+      // The assumption is that the display set at position 0 is the first
+      // study being displayed, and is thus the "active" study.
+      const studyMap = {};
+      const studies = displaySets.
+        reduce((prev, curr) => {
+          const { StudyInstanceUID } = curr;
+          if (!studyMap[StudyInstanceUID]) {
+            const study = DicomMetadataStore.getStudy(StudyInstanceUID);
+            studyMap[StudyInstanceUID] = study;
+            prev.push(study);
+          }
+          return prev;
+        }, []);
```

Comment:
So we are creating studies object here again and again, when DS are being added? Why there is no garantee that `.getStudy` gives us the study metadata? since later you just use the first study `const studyMetadata = studies[0]` 

Also here we are running the HP engine on EVERY displaySet creation, is that right? Does that mean that every time there is a new ranking (since DS are in the process of creation as we match).

Reply:
There was existing code that built up the list of studies in the HangingProtocolService itself, assuming that the run method got called multiple times, as each series got added.  However, it was broken because the order of the calls HAD to match the order of the series being matched or else it would fail to display the right series (eg the correct match).  That allowed the Pet SUV table to work as expected IF the basic match worked, but would fail to display anything if the first series found was incorrect.    So here, what I'm trying to do is make it work all at once, with a single DisplaySets changed event.  The change event only gets called once per study creation, right at the end - everything else is held until that point (see the holdChangeEvents code in DisplaySetService).   Then, since the run method needs the list of studies, the list is created externally and passed in - since we don't want to depend on the metadata service.
I'm trying to make this not have race conditions while minimizing how many changes are made.